pkg/app/ops/firestoreindexcreator/indexcreator.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package firestoreindexcreator automatically creates the needed composite indexes for Google Firestore,
+// based on well-defined JSON format indexes list with the name indexes.json.
+package firestoreindexcreator
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+)
+
+type IndexCreator interface {
+	CreateIndexes(ctx context.Context) error
+}
+
+type firestoreClient interface {
+	authorize(ctx context.Context) error
+	createIndex(ctx context.Context, idx *index)
+}
+
+type indexCreator struct {
+	firestoreClient
+	logger *zap.Logger
+}
+
+func NewIndexCreator(gcloudPath, projectID, serviceAcccountFile string, logger *zap.Logger) IndexCreator {
+	return &indexCreator{
+		// TODO: Use Go SDK to create Firebase indexes upon it will be started providing
+		firestoreClient: newGcloud(gcloudPath, projectID, serviceAcccountFile, logger),
+		logger:          logger.Named("firestore-index-creator"),
+	}
+}
+
+// CreateIndexes creates needed composite indexes for Google Firestore based on
+// well-defined indexes list.
+func (c *indexCreator) CreateIndexes(ctx context.Context) error {
+	if err := c.authorize(ctx); err != nil {
+		return fmt.Errorf("failed to authorize: %w", err)
+	}
+
+	idxs, err := parseIndexes()
+	if err != nil {
+		return err
+	}
+	for i := 0; i < len(idxs.Indexes); i++ {
+		// TODO: Check if the index is already added and if so skip creating
+		c.createIndex(ctx, &idxs.Indexes[i])
+	}
+	return nil
```

Comment:
It will be nice if we have some log entries to show the process 
- start checking and ensuring the existence of %d indexes
- completed checking all indexes // and add a field to show the elapsed time.

Reply:
That's the case. Ideally, I'd also like to track the progress of indexes creating and ensure the operation in progress to complete. But for that, seems like additional permission is required. (I mean, it fails when starting to watch the progress because of permission denied).

And honestly, I'm not sure exactly what permission is needed because I thought it's okay that we don't have such log so far.

Reply:
Can I just leave a TODO comment to make it a future task, or work on it right now?

Reply:
FYI: That's why I settled on running in the async mode which returns immediately without waiting for the operation in progress to complete.
https://github.com/pipe-cd/pipe/pull/1553/files#diff-e0a87ea7ae118e135da5409141a36371cc95870dc3f2985ce1c9663e029455a7R73

Reply:
> But for that, seems like additional permission is required. (I mean, it fails when starting to watch the progress because of permission denied).

Sorry, I don't get this part. What did you mean?
I just wanted to say that we should have a log message before the loop and one after the loop to show that the task will be started and it was completed after x seconds.

Reply:
Okay, it looks good 👍  But I'm wondering if it's really meaningful? I mean, the gcloud command will terminate without waiting for the index creation to complete, so the task will be completed immediately, because of `--async` option.

Forgive me that rough sentence, I'm soon-to-be in a MTG. I don't intend to hurt you 🙏

Reply:
>Sorry, I don't get this part.

To watch the progress logs, we need additional permission aside from the `Firestore author` role.

Reply:
> I mean, the gcloud command will terminate without waiting for the index creation to complete, so the task will be completed immediately, because of --async option.

Nice catch! I thought it will be waiting for the completion.
Sorry, I did not notice the `async` flag.

>  I don't intend to hurt you

lol. That is ok. 👍